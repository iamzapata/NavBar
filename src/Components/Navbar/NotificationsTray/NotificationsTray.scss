@import "styles/variables.scss";

.NotificationsTray {
  a {
    position: relative;
    border-radius: 2rem;
    height: 100%;
    width: 45px;
    border-color: $color-transparent-gray-border;
    box-shadow: 0px 1px 0px $color-gray-border;
  }

  .Notifications__Count {
    background: #ca2127;
    position: absolute;
    color: white;
    border-radius: 1rem;
    width: 21px;
    height: 16px;
    font-size: 0.7rem;
    right: -10px;
    bottom: 20px;
  }
}
