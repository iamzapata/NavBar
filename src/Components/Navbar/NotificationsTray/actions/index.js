import { request, sleep } from "utils/request"
import ActionTypes from "../actionTypes"

const getNotificationsRequest = () => {
  return {
    type: ActionTypes.GET_NOTIFICATIONS_REQUEST,
    isLoading: true
  }
}

const getNotificationsSuccess = notifications => {
  return {
    type: ActionTypes.GET_NOTIFICATIONS_SUCCESS,
    isLoading: false,
    notifications
  }
}

const getNotificationsFailure = err => {
  return {
    type: ActionTypes.GET_NOTIFICATIONS_FAILURE,
    isLoading: false,
    err
  }
}

export default function getNotifications() {
  return async dispatch => {
    dispatch(getNotificationsRequest())
    try {
      await sleep(500)
      const notifications = await request("notifications/index.json")
      dispatch(getNotificationsSuccess(notifications))
    } catch (err) {
      dispatch(getNotificationsFailure(err))
    }
  }
}
