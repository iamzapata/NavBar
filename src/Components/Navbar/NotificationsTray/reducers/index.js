import { createSelector } from "reselect"
import { order } from "utils/dates"
import ActionTypes from "../actionTypes"

const defaultState = {
  notifications: [],
  isLoading: false,
  err: null
}

export default function NotificationsStore(state = defaultState, action) {
  const { type, notifications, isLoading, err } = action
  switch (type) {
    case ActionTypes.GET_NOTIFICATIONS_REQUEST:
      return {
        ...state,
        isLoading
      }
    case ActionTypes.GET_NOTIFICATIONS_SUCCESS:
      return {
        ...state,
        notifications,
        isLoading
      }
    case ActionTypes.GET_NOTIFICATIONS_FAILURE:
      return {
        ...state,
        isLoading,
        err
      }
    default:
      return state
  }
}

const getNotifications = state => state.Notifications.notifications

const getNotificationsLoading = state => state.Notifications.isLoading

export const notificationsSelector = createSelector(
  [getNotifications],
  notifications => {
    return notifications.sort((a, b) => order(a, b)("updated_at")("desc"))
  }
)

export const notificationsLoadingSelector = createSelector(
  [getNotificationsLoading],
  notificationsLoading => notificationsLoading
)
