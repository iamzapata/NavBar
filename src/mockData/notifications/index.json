[
  {
    "id": 1,
    "content": "Suspendisse potenti. In eleifend quam a odio.",
    "created_at": "2018-04-05 01:13:44",
    "type": "feed",
    "read": false
  },
  {
    "id": 2,
    "content": "Proin at turpis a pede posuere nonummy. Integer non velit.",
    "created_at": "2018-07-16 14:18:19",
    "type": "feed",
    "read": false
  },
  {
    "id": 3,
    "content": "Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.",
    "created_at": "2018-02-19 12:18:28",
    "type": "feed_activity",
    "read": true
  },
  {
    "id": 4,
    "content": "Duis bibendum, felis sed interdum venenatis, turpis enim blandit mi, in porttitor pede justo eu massa. Donec dapibus.",
    "created_at": "2018-01-23 10:21:28",
    "type": "feed",
    "read": false
  },
  {
    "id": 5,
    "content": "Nulla suscipit ligula in lacus.",
    "created_at": "2018-01-08 00:43:33",
    "type": "feed_activity",
    "read": true
  },
  {
    "id": 6,
    "content": "Morbi vestibulum, velit id pretium iaculis, diam erat fermentum justo, nec condimentum neque sapien placerat ante.",
    "created_at": "2018-05-29 04:25:41",
    "type": "feed_activity",
    "read": false
  },
  {
    "id": 7,
    "content": "In sagittis dui vel nisl. Duis ac nibh.",
    "created_at": "2018-05-16 09:48:45",
    "type": "feed",
    "read": false
  }
]
