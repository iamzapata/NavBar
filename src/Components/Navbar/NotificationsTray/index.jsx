import React from "react"
import { arrayOf, shape, number, string } from "prop-types"
import "./NotificationsTray.scss"

const NotificationsTray = ({ notifications }) => (
  <div className="NotificationsTray navbar-item">
    <a className="button">
      <i className="fa fa-bell" />
      {notifications.length > 0 && (
        <span className="Notifications__Count">{notifications.length} +</span>
      )}
    </a>
  </div>
)

NotificationsTray.propTypes = {
  notifications: arrayOf(
    shape({
      id: number.isRequired,
      content: string.isRequired,
      created_at: string.isReqired,
      type: string.isRequired
    })
  ).isRequired
}

export default NotificationsTray
