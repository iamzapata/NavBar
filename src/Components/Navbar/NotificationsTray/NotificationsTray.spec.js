import React from "react"
import { render, cleanup, fireEvent } from "react-testing-library"
import NotificationsTray from "./index"
import notifications from "mockData/notifications/index.js"
import "jest-dom/extend-expect"

describe("<NotificationsTray />", () => {
  afterEach(cleanup)
  it("Should render self", () => {
    const { container } = render(
      <NotificationsTray notifications={notifications} />
    )
    expect(container.firstChild).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it("Should show the number of notifications", () => {
    const { container, queryByText, debug } = render(
      <NotificationsTray notifications={notifications} />
    )
    expect(container).toHaveTextContent("7")
  })
})
