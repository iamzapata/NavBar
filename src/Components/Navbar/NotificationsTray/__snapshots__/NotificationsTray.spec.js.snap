// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsTray /> Should render self 1`] = `
<div
  class="NotificationsTray navbar-item"
>
  <a
    class="button"
  >
    <i
      class="fa fa-bell"
    />
    <span
      class="Notifications__Count"
    >
      7
       +
    </span>
  </a>
</div>
`;
